micropython/lora/lora/sx1276.py

```diff
@@ -0,0 +1,592 @@
+_MODEM_CONFIG1_BW10_4 = const(0x1)
+_MODEM_CONFIG1_BW15_6 = const(0x2)
+_MODEM_CONFIG1_BW20_8 = const(0x3)
+_MODEM_CONFIG1_BW31_25 = const(0x4)
+_MODEM_CONFIG1_BW41_7 = const(0x5)
+_MODEM_CONFIG1_BW62_5 = const(0x6)
+_MODEM_CONFIG1_BW125 = const(0x7)
+_MODEM_CONFIG1_BW250 = const(0x8)  # not supported in lower band (169MHz)
+_MODEM_CONFIG1_BW500 = const(0x9)  # not supported in lower band (169MHz)
+_MODEM_CONFIG1_CODING_RATE_MASK = const(0x7)
+_MODEM_CONFIG1_CODING_RATE_SHIFT = const(1)
+_MODEM_CONFIG1_CODING_RATE_45 = const(0b001)
+_MODEM_CONFIG1_CODING_RATE_46 = const(0b010)
+_MODEM_CONFIG1_CODING_RATE_47 = const(0b011)
+_MODEM_CONFIG1_CODING_RATE_48 = const(0b100)
+_MODEM_CONFIG1_IMPLICIT_HEADER_MODE_ON = const(1 << 0)
+
+_REG_MODEM_CONFIG2 = const(0x1E)
+_MODEM_CONFIG2_SF_MASK = const(0xF)  # Spreading Factor
+_MODEM_CONFIG2_SF_SHIFT = const(4)
+# SF values are integers 6-12 for SF6-SF12, so skipping constants for these
+_MODEM_CONFIG2_SF_MIN = const(6)  # inclusive
+_MODEM_CONFIG2_SF_MAX = const(12)  # inclusive
+
+_MODEM_CONFIG2_TX_CONTINUOUS = const(1 << 3)
+_MODEM_CONFIG2_RX_PAYLOAD_CRC_ON = const(1 << 2)
+_MODEM_CONFIG2_SYMB_TIMEOUT_MSB_MASK = 0x3
+
+_REG_SYMB_TIMEOUT_LSB = const(0x1F)
+
+_REG_PREAMBLE_LEN_MSB = const(0x20)
+_REG_PREAMBLE_LEN_LSB = const(0x21)
+
+_REG_PAYLOAD_LEN = const(0x22)  # Only for implicit header mode
+_REG_MAX_PAYLOAD_LEN = const(0x23)
+
+_REG_HOP_PERIOD = const(0x24)
+
+_REG_FIFO_TXBYTE_ADDR = const(0x25)
+
+_REG_MODEM_CONFIG3 = const(0x26)
+_MODEM_CONFIG3_AGC_ON = const(1 << 2)
+_MODEM_CONFIG3_LOW_DATA_RATE_OPTIMIZE = const(1 << 3)
+
+_REG_DETECT_OPTIMIZE = const(0x31)
+_DETECT_OPTIMIZE_AUTOMATIC_IF_ON = const(
+    1 << 7
+)  # Bit should be cleared after reset, as per errata
+_DETECT_OPTIMIZE_MASK = 0x7
+_DETECT_OPTIMIZE_SF6 = const(0x05)
+_DETECT_OPTIMIZE_OTHER = const(0x03)
+
+_REG_INVERT_IQ = const(0x33)
+_INVERT_IQ_RX = const(1 << 6)
+_INVERT_IQ_TX = const(1 << 0)
+
+_REG_DETECTION_THRESHOLD = const(0x37)
+_DETECTION_THRESHOLD_SF6 = const(0x0C)
+_DETECTION_THRESHOLD_OTHER = const(0x0A)  # SF7 to SF12
+
+_REG_SYNC_WORD = const(0x39)
+
+_REG_FSKOOK_IMAGE_CAL = const(0x3B)  # NOTE: Only accessible in FSK/OOK mode
+_IMAGE_CAL_START = const(1 << 6)
+_IMAGE_CAL_RUNNING = const(1 << 5)
+_IMAGE_CAL_AUTO = const(1 << 7)
+
+_REG_DIO_MAPPING1 = const(0x40)
+_DIO0_MAPPING_MASK = const(0x3)
+_DIO0_MAPPING_SHIFT = const(6)
+_DIO1_MAPPING_MASK = const(0x3)
+_DIO1_MAPPING_SHIFT = const(4)
+_DIO2_MAPPING_MASK = const(0x3)
+_DIO2_MAPPING_SHIFT = const(2)
+_DIO3_MAPPING_MASK = const(0x3)
+_DIO3_MAPPING_SHIFT = const(0)
+
+_REG_DIO_MAPPING2 = const(0x41)
+_DIO4_MAPPING_MASK = const(0x3)
+_DIO4_MAPPING_SHIFT = const(6)
+_DIO5_MAPPING_MASK = const(0x3)
+_DIO5_MAPPING_SHIFT = const(4)
+
+_REG_VERSION = const(0x42)
+
+
+def _clamp(v, vmin, vmax):
+    return min(max(vmin, v), vmax)
+
+
+class SX1276(BaseModem):
+    def __init__(self, spi, cs, dio0=None, dio1=None, radio_reset=None, lora_cfg=None):
```

Comment:
The fields for lora_cfg definitely need to be fully documented!

There can't really be a default because there isn't a single frequency that works everywhere in the world, although that minimal config `{ 'freq_khz': xyz }` should give a working radio in many cases (but recommend to also set `sf`, `bw`, and probably other settings!).

Reply:
Documented now.